### DISCUSSION TITLE
Many installed versions wrong

### DISCUSSION THREAD

**Discussion by OLLI-S** [microsoft/winget-pkgs#57695] [General]
Winget reports me some updates:
![image](https://user-images.githubusercontent.com/8037790/163266189-54e18948-9614-4cf9-8518-2c5dfc1bbf95.png)

For all these apps (marked with a red border) the installed version is not correct (also not in Windows 10).
When I open these apps, I see, that the installed version is matching the available version (that I already have installed the correct versions).
I guess that these apps do not update the version number in Windows, after they update themselves (when I start them they update themselves).

All apps are installed manually (downloaded and installed).
I did not yet use Winget to install any package.

Can this issue be solved by Winget, or must the version number be entered correctly by the app itself?

**Comment by OfficialEsco** [comment]
Thats... weird
* Discord - I've seen Discord being weird with updates, but its usually resolved by running the Installer
* TeamSpeak - This makes no sense why its displaying the wrong version
* Ubisoft Connect - Its displaying 85.1 which is SUUUUPER old, also makes no sense why its not updating as it does for me every time..
* Epic Games Launcher - This is a MSI so i believe it ALWAYS updates the DisplayVersion
* GOG Galaxy - This one is missing a DisplayVersion (the community have complained and is waiting for a fix), in winget version 1.3 these will be hidden and needs to use --include-unknown for it to be updated

Overall this seems to be a issue on your end, possibly a change in the registry path so you have 2 entries but the oldest one is first(?)
If you've already done `winget upgrade <ID>` i suggest you to reinstall the Applications, if you want to be on the safe side or just monitor the registry/left over files i'd suggest you to use Revo Uninstaller.

**Comment by OLLI-S** [comment]
Some more info:
* When I start **TeamSpeak** and go to Help -> About, I see version **3.5.6** (so the current version is installed)
* When I launch **Ubisoft Connect** then there is no update installed, I can log into the app (if there is an update available, then it is automatically installed before I can log in)
* When I start **Discord**, then there are no updates installed, because I use Discord very often
* I started **Epic Games Launcher** and there was an update available and installed. After the update was installed, Winget still reported an update (checked for updates via `winget upgrade` several times, over a period of 30 minutes).

So I manually updated all listed apps (via `winget upgrade name`).
This was successful for all applications except:
* **GOG GALAXY** here the installer told me that the actual version is already installed, so the installation will be cancelled)
* **Microsoft.VC++2015-2019Redist-x64** Update starts and closed immediately
Same if I do `winget upgrade --all`

I also intalled **USB Safely Remove** via Winget and this package is now also listed in the available updates.
Reason: the installed version is "unknown".

**Reply by OfficialEsco** [reply]
the TeamSpeak versions are far apart (1 year), i have not used TeamSpeak in 10+ years so i'm unsure if they got a auto updater? if so maybe it bugged? or the manual upgrade you did messed it up?

Yeah Ubisoft Connect updates way before you can do anything, and as i've used WinGet instead of opening the program i know it does set the DisplayVersion every time, so its weird that it has not

Discord is funky, the previous version of Discord does not get uninstalled, it just creates a new folder at `%LOCALAPPDATA%\Discord`, so they installer might not work at times, i've also noticed BetterDiscord causes issues with the DisplayVersion string

We've had so many issues with VC++... I'm not sure if the 1.2 client resolves the issue, but i know 1.3 does, the main issue here is that its mixing the x64 and x86 installation

Any software with the version Unknown is caused by the Developer not setting the DisplayVersion string on Installation/Update, if you see these you should complain to the Developer, but as i said we have a workaround for this in the 1.3 winget client.

Also i've seen some softwares update their DisplayVersion after closing the software

**Reply by OLLI-S** [reply]
I installed **TeamSpeak** many many years ago and also updated it via the auto updater.
Started using WinGet 2 days ago. :-)

I suppose it is the best to install all new apps via WinGet (as far as they are available) and also update them only via WinGet.
So WinGet has the change to update them, before the auto updater is doing this....

But I recognized, that for some apps there are updates available via the auto updater, but in WinGet there is no update avaiable.
I suppose the WinGet packages are maintained by the community, and not by the vendor of these apps, right?

**Reply by OfficialEsco** [reply]
Most of the time WinGet just reruns the installer with silent switches unless we tell it otherwise, which in theory can be alot better than most auto updaters for some reason..

The WinGet-Pkgs source is maintained by the Community, however it is mainly maintained by 5+ people...

* GitHub releases are not a problem since it got a nice RSS feed, they can also be easily automated with a community made bot [winget-pkgs-automation-bot](https://github.com/apps/winget-pkgs-automation-bot) which is currently maintained [here](https://github.com/vedantmgoyal2009/vedantmgoyal2009/tree/main/winget-pkgs-automation) until Microsoft imports it.

* SourceForge releases are also not a problem, however maybe a bit more tricky to automate

* Now the real issue comes down to Packages which does not have a RSS feed or a API which the automation can use, the few main package maintainers cannot monitor every non GitHub/SourceForge package, this has to be done by someone who uses the software

The MSStore source is maintained by the actual vendor

* However this also correlates to the actual Microsoft Store

* There are a bunch of Manifest issues because they have less experience and we with experience cannot help as its a different private repository.

We are actually waiting for real Vendors to control a Package inside this Repository, but this takes time since the feature is not ready yet.

**Comment by OLLI-S** [comment]
### Fantastic Job
First of all: you all are doing here a fantastic job!
Thank you, WinGet is really a powerful and helpful tool.

### WinGet Results
At my PC I have 3 apps left, that cause problems:

```
Name                                               ID                                Version       Verfügbar     Quelle
-----------------------------------------------------------------------------------------------------------------------
USB Safely Remove                                  CrystalRich.USBSafelyRemove       Unknown       6.4.2.1297    winget
GOG GALAXY                                         GOG.Galaxy                        Unknown       2.0.46.133    winget
Microsoft Visual C++ 2015-2019 Redistributable (x… Microsoft.VC++2015-2019Redist-x64 14.28.29334.0 14.29.30139.0 winget
3 Aktualisierungen verfügbar.
```
**Microsoft Visual C++ 2015-2019 Redistributable**
Typing ```winget upgrade Microsoft.VC++2015-2019Redist-x64``` tells me ```No applicable updates were found.```
I get a German message and I translated it, so I hope my translation is correct.

**GOG GALAXY and USB Safely Remove**
They have an ```Unknown``` Version, so how can I help you solving this problem?
Should I contact the vendors of these apps?
I am happy to help!

**Reply by lychichem** [reply]
For software with unknown version numbers, YES, you have to contact the developer of the software to ask them adding the behavior to write version number in registry. Winget only read these info from registry.

**Reply by OLLI-S** [reply]
@lychichem Do you have any more details for the 3ed party companies, what keys should be written?
The path in the registry and the name of the key containing the value.
Thank you!

**Reply by OfficialEsco** [reply]
It depends if they install it in the User or Machine scope
For Machine `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\<ProductCode>`
For CurrentUser `HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Uninstall\<ProductCode>`
The most important strings are `Publisher`, `DisplayName` and `DisplayVersion`.

The Installer usually have a similar code to this that adds those keys
`WriteRegStr HKLM "Software\Microsoft\Windows\CurrentVersion\Uninstall\<ProductCode>" "DisplayVersion" "${VERSION}"`

I'm not familiar with Winget-CLI but it probably also looks for the `UninstallString` and `QuietUninstallString` with the Uninstall command.
